File: quote/Projects/PluginModel/DetailSortComparer.vb
Imports System.ComponentModel
Imports System.Globalization

Namespace Common
    Public Class DetailSortComparer
        Implements IComparer(Of Detail)
        Private _PropDesc As PropertyDescriptor = Nothing
        Private _Direction As ListSortDirection = ListSortDirection.Ascending
        Public Sub New(ByVal propDescriptor As PropertyDescriptor, ByVal direction As ListSortDirection)
            Me._PropDesc = propDescriptor
            Me._Direction = direction
        End Sub
        Public Function Compare(ByVal x As Detail, ByVal y As Detail) As Integer Implements IComparer(Of Detail).Compare
            Dim retValue As Integer
            Dim xValue As Object = _PropDesc.GetValue(x)
            Dim yValue As Object = _PropDesc.GetValue(y)
            retValue = CompareValues(xValue, yValue, _Direction)
            Return retValue
        End Function

        Private Shared Function CompareValues(ByVal x As Object, ByVal y As Object, ByVal direction As ListSortDirection) As Integer
            Dim retValue As Integer = 0
            If IsNumeric(x) And IsNumeric(y) Then
                retValue = CType(x, Double).CompareTo(CType(y, Double))
            ElseIf x Is Nothing Then
                retValue = 0
            ElseIf y Is Nothing Then
                retValue = 0
            Else
                retValue = String.Compare(x.ToString(), y.ToString(), True, CultureInfo.CurrentCulture)
            End If
            If direction = ListSortDirection.Descending Then
                retValue *= -1
            End If
            Return retValue
        End Function

    End Class


End Namespace

File: quote/Projects/PluginModel/DetailSortComparer.cs
namespace Model.Common
{
    using System;
    using System.Collections.Generic;
    using System.ComponentModel;
    using System.Globalization;
    using System.Linq;

    public class DetailSortComparer : IComparer<Detail>
    {
        private PropertyDescriptor propDesc = null;
        
        private ListSortDirection direction = ListSortDirection.Ascending;
        
        public DetailSortComparer(PropertyDescriptor propDescriptor, ListSortDirection direction)
        {
            this.propDesc = propDescriptor;
            this.direction = direction;
        }
        
        public int Compare(Detail x, Detail y)
        {
            int retValue = 0;
            object leftValue = this.propDesc.GetValue(x);
            object rightValue = this.propDesc.GetValue(y);
            retValue = CompareValues(leftValue, rightValue, this.direction);
            return retValue;
        }
        
        private static bool IsNumeric(string value)
        {
            int result;
            if (int.TryParse(value, out result))
            {
                return true;
            }
            
            return false;
        }

        private static int CompareValues(object x, object y, ListSortDirection direction)
        {
            int retValue = 0;
            
            if (IsNumeric(x.ToString()) & IsNumeric(y.ToString())) 
            {
                double left = Convert.ToDouble(x, CultureInfo.CurrentCulture);
                double right = Convert.ToDouble(y, CultureInfo.CurrentCulture);
                
                retValue = left.CompareTo(right);
            } 
            else if (x == null) 
            {
                retValue = 0;
            } 
            else if (y == null) 
            {
                retValue = 0;
            } 
            else 
            {
                retValue = string.Compare(x.ToString(), y.ToString(), true, CultureInfo.CurrentCulture);
            }
            
            if (direction == ListSortDirection.Descending) 
            {
                retValue *= -1;
            }
            
            return retValue;
        }
    }
}